ethcontract/src/batch.rs

```diff
@@ -0,0 +1,89 @@
+//! Module containing components to batch multiple contract calls
+//! into a single request to the Node.
+
+use futures::channel::oneshot::{channel, Sender};
+use web3::{
+    error::Error as Web3Error,
+    helpers::{self},
+    types::{BlockId, BlockNumber, Bytes, CallRequest},
+    BatchTransport as Web3BatchTransport,
+};
+
+/// Struct allowing to batch multiple calls into a single Node request
+pub struct CallBatch<T: Web3BatchTransport> {
+    inner: T,
+    requests: Vec<(Request, CompletionHandler)>,
+}
+
+type Request = (CallRequest, Option<BlockId>);
+type CompletionHandler = Sender<Result<Bytes, Web3Error>>;
+
+impl<T: Web3BatchTransport> CallBatch<T> {
+    /// Create a new instance from a BatchTransport
+    pub fn new(inner: T) -> Self {
+        Self {
+            inner,
+            requests: Default::default(),
+        }
+    }
+
+    /// Adds a call request to the current batch. The resulting future can only resolve after
+    /// the batch has been resolved via `execute_all`.
+    /// Panics, if the batch is dropped before executing.
+    pub fn push(
+        &mut self,
+        call: CallRequest,
+        block: Option<BlockId>,
+    ) -> impl std::future::Future<Output = Result<Bytes, Web3Error>> {
+        let (tx, rx) = channel();
+        self.requests.push(((call, block), tx));
+        async move { rx.await.expect("Batch has been dropped without executing") }
+    }
+
+    /// Execute and resolve all enqueued CallRequests in a single RPC call
+    pub async fn execute_all(self) -> Result<(), Web3Error> {
+        let results = self
+            .inner
+            .send_batch(self.requests.iter().map(|((request, block), _)| {
+                let req = helpers::serialize(request);
+                let block =
+                    helpers::serialize(&block.unwrap_or_else(|| BlockNumber::Latest.into()));
+                let (id, request) = self.inner.prepare("eth_call", vec![req, block]);
+                (id, request)
+            }))
+            .await?;
+        for (result, (_, sender)) in results.into_iter().zip(self.requests.into_iter()) {
+            sender.send(result.and_then(helpers::decode)).unwrap();
```

Comment:
Sending on the channel fails if the receiver has been dropped. Even though it is likely a programmer error, it can be reasonable to drop a future without awaiting it, so we shouldn't panic in this case.

Reply:
Totally agree here. I must have changed it when I was debugging an issue but was meaning to ignore errors here.

Reply:
I also don't think failing to send on the result channel should panic. While I can't think of a valid use case, I don't see why dropping the request future should be disallowed.

Maybe a good compromise is to log a warning in this case, but otherwise ignore the error.